iceoryx_dds/include/iceoryx_dds/gateway/dds_gateway_generic.hpp

```diff
@@ -0,0 +1,135 @@
+
+  public:
+    virtual ~DDSGatewayGeneric() noexcept;
+
+    DDSGatewayGeneric(const DDSGatewayGeneric&) = delete;
+    DDSGatewayGeneric& operator=(const DDSGatewayGeneric&) = delete;
+    DDSGatewayGeneric(DDSGatewayGeneric&&) = delete;
+    DDSGatewayGeneric& operator=(DDSGatewayGeneric&&) = delete;
+
+    void runMultithreaded() noexcept;
+    void shutdown() noexcept;
+
+    ///
+    /// @brief loadConfiguration Load the provided configuration.
+    /// @param config Generic dds gateway configuration which is applicable to all implementations.
+    ///
+    virtual void loadConfiguration(GatewayConfig config) noexcept = 0;
+    ///
+    /// @brief discover Process discovery messages coming from iceoryx.
+    /// @param msg The discovery message.
+    ///
+    virtual void discover(const iox::capro::CaproMessage& msg) noexcept = 0;
+    ///
+    /// @brief forward Forward data between the two terminals of the channel used by the implementation.
+    /// @param channel The channel to propogate data across.
+    ///
+    virtual void forward(channel_t channel) noexcept = 0;
+
+    uint64_t getNumberOfChannels() const noexcept;
+
+  protected:
+    DDSGatewayGeneric() noexcept;
+
+    ChannelFactory m_channelFactory;
+
+    ///
+    /// @brief addChannel Creates a channel for the given service and stores a copy of it in an internal collection for
+    /// later access.
+    /// @param service The service to create a channel for.
+    /// @return A copy of the created channel.
+    ///
+    /// @note Channels are supposed to be lightweight, consisting only of pointers to the terminals and a copy of the
+    /// service description, therefore a copy is provided to any entity that requires them.
+    /// When no more copies of a channel exists in the system, the terminals will automatically be cleaned up via
+    /// the custom deleters included in their pointers.
+    ///
+    /// The service description is perhaps too large for copying since they contain strings, however this should be
+    /// addressed with a service description repository feature.
+    ///
+    channel_t addChannel(const iox::capro::ServiceDescription& service) noexcept;
+
+    ///
+    /// @brief findChannel Searches for a channel for the given service in the internally stored collection and returns
+    /// it one exists.
+    /// @param service The service to find a channel for.
+    /// @return An optional containining the matching channel if one exists, otherwise an empty optional.
+    ///
+    iox::cxx::optional<channel_t> findChannel(const iox::capro::ServiceDescription& service) const noexcept;
+
+    ///
+    /// @brief forEachChannel Executs the given function for each channel in the internally stored collection.
+    /// @param f The function to execute.
+    /// @note This operation allows thread-safe access to the internal collection.
+    ///
+    void forEachChannel(const std::function<void(channel_t&)> f) const noexcept;
```

Comment:
Please do not any new `std::function` usage, we're removing it with #131 

I this case you should be easily able to replace it with `cxx::function_ref` as it is used as a function parameter.

Reply:
:heavy_check_mark:

Reply:
A note on using `cxx::function_ref`: as discussed, this can only be done here. 
The other instances where `std::function` is being used are factory methods which are being stored in class instances. These will remain as `std::function`, and like with smart pointers, will be looked at more closely when it comes to certification since the solution is not so straight forward.